src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1734,99 +1734,139 @@ private BoundPattern BindBinaryPattern(
+            BoundPattern? result = null;
 
-                static void collectCandidates(BoundPattern pat, ArrayBuilder<TypeSymbol> candidates)
+            while (binaryPatternStack.TryPop(out var binaryPatternAndPermitDesignations))
+            {
+                (var binaryPattern, permitDesignations) = binaryPatternAndPermitDesignations;
+                if (result == null)
                 {
-                    if (pat is BoundBinaryPattern { Disjunction: true } p)
-                    {
-                        collectCandidates(p.Left, candidates);
-                        collectCandidates(p.Right, candidates);
-                    }
-                    else
-                    {
-                        candidates.Add(pat.NarrowedType);
-                    }
+                    Debug.Assert(binaryPattern.Left is not BinaryPatternSyntax);
+                    result = BindPattern(binaryPattern.Left, inputType, permitDesignations, hasErrors, diagnostics);
                 }
 
-                TypeSymbol? leastSpecificType(SyntaxNode node, ArrayBuilder<TypeSymbol> candidates, BindingDiagnosticBag diagnostics)
+                result = bindBinaryPattern(result, this, binaryPattern, inputType, permitDesignations, hasErrors, diagnostics);
+            }
+
+            binaryPatternStack.Free();
+            Debug.Assert(result != null);
+            return result;
+
+            static BoundPattern bindBinaryPattern(
+                BoundPattern preboundLeft,
+                Binder binder,
+                BinaryPatternSyntax node,
+                TypeSymbol inputType,
+                bool permitDesignations,
+                bool hasErrors,
+                BindingDiagnosticBag diagnostics)
+            {
+                bool isDisjunction = node.Kind() == SyntaxKind.OrPattern;
+                if (isDisjunction)
                 {
-                    Debug.Assert(candidates.Count >= 2);
-                    CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
-                    TypeSymbol? bestSoFar = candidates[0];
-                    // first pass: select a candidate for which no other has been shown to be an improvement.
-                    for (int i = 1, n = candidates.Count; i < n; i++)
+                    Debug.Assert(!permitDesignations);
+                    MessageID.IDS_FeatureOrPattern.CheckFeatureAvailability(diagnostics, node.OperatorToken);
+
+                    var right = binder.BindPattern(node.Right, inputType, permitDesignations, hasErrors, diagnostics);
+
+                    // Compute the common type. This algorithm is quadratic, but disjunctive patterns are unlikely to be huge
+                    var narrowedTypeCandidates = ArrayBuilder<TypeSymbol>.GetInstance(2);
+                    collectCandidates(preboundLeft, narrowedTypeCandidates);
```

Comment:
>collectCandidates(preboundLeft, narrowedTypeCandidates); [](http://example.com/codeflow?start=20&length=56)

Perhaps the assumption from the comment above that "disjunctive patterns are unlikely to be huge" has been proven wrong. Therefore, it might make sense keep the set of candidate types from the left as we bubble up, instead of keep revisiting the subtree over and over again. Also, it might be worth considering to collect only distinct types to decrease amount of memory used.

Reply:
I'll take the former suggestion, but I'll avoid the latter suggestion for now. We can do it in a followup if it's necessary, but it's a bigger change and I don't think we need to look at it yet.